cassiopeia/tools/lineage_tracing_simulator.py

```diff
@@ -0,0 +1,78 @@
+import abc
+
+import numpy as np
+
+from .tree import Tree
+
+
+class LineageTracingSimulator(abc.ABC):
+    r"""
+    Abstract base class for all lineage tracing simulators.
+
+    A LineageTracingSimulator implements a method overlay_lineage_tracing_data
+    which overlays lineage tracing data (i.e. character vectors) on top of the
+    tree. These are stored as the node's state.
+    """
+
+    @abc.abstractmethod
+    def overlay_lineage_tracing_data(self, tree: Tree) -> None:
+        r"""
+        Annotates the tree's nodes with lineage tracing character vectors.
+        These are stored as the node's state. (Operates on the tree in-place.)
+
+        Args:
+            tree: The tree to overlay lineage tracing data on.
+        """
+
+
+class IIDExponentialLineageTracer(LineageTracingSimulator):
+    r"""
+    Characters evolve IID over the lineage, with the same given mutation rate.
+
+    Args:
+        mutation_rate: The mutation rate of each character (same for all).
+        num_characters: The number of characters.
+    """
+
+    def __init__(self, mutation_rate: float, num_characters: float):
+        self.mutation_rate = mutation_rate
+        self.num_characters = num_characters
+
+    def overlay_lineage_tracing_data(self, tree: Tree) -> None:
+        r"""
+        See base class.
+        """
+        num_characters = self.num_characters
+        mutation_rate = self.mutation_rate
+
+        def dfs(node: int, tree: Tree):
+            node_state = tree.get_state(node)
+            for child in tree.children(node):
+                # Compute the state of the child
+                child_state = ""
+                edge_length = tree.get_age(node) - tree.get_age(child)
+                # print(f"{node} -> {child}, length {edge_length}")
+                assert edge_length >= 0
+                for i in range(num_characters):
+                    # See what happens to character i
+                    if node_state[i] != "0":
+                        # The character has already mutated; there in nothing
+                        # to do
+                        child_state += node_state[i]
+                        continue
+                    else:
+                        # Determine if the character will mutate.
+                        mutates = (
+                            np.random.exponential(1.0 / mutation_rate)
+                            < edge_length
```

Comment:
I could be misremembering, but I believe `np.random.exponential` already inverts the scale parameter (i.e. `np.random.exponential(b)` draws from `1/b * exp(x/b)`). [You can read the docs here: https://numpy.org/doc/stable/reference/random/generated/numpy.random.exponential.html -- indeed, it looks like the function inverts the scale parameter for you]

If this is the case, do you want to pass instead use `np.random.exponential(mutation_rate)`?

Reply:
`np.random.exponential` uses the scale parameterization, so we should invert the rate as in `np.random.exponential(1.0 / mutation_rate)`, right?

Reply:
Sorry - I'm a bit confused. If the density function you want is  `lambda * exp(x*lambda)` (where `lambda` is the mutation rate) then you are good to go.